### PR TITLE
♻️Introduce waiting for body into DomWriter

### PR DESCRIPTION
**Pull request by calebcordry**
Working on using DomWriter in a4a, however ads need to fetch ASAP (layoutMeasure) but delay any rendering until layoutCallback.

This change makes DomWriter wait for a promise of a body to write to, before doing any transferring of elements. Open to other ideas of how to make this work.

### CODE REVIEW

File: src/utils/dom-writer.js [hunk 1]
```diff
@@ -178,31 +178,33 @@ export class DomWriterStreamer {
...
+          transferCount++;
+          targetBody.appendChild(inputBody.firstChild);
+        }
+        if (transferCount > 0) {
+          this.onBodyChunk_();
+        }
+
+        // EOF.
+        if (this.eof_) {
+          this.onEnd_();
```
Comment (calebcordry) [inline src/utils/dom-writer.js]: Wanted to clarify that the intention here is for this to be called after DOM has been transferred? The jsdoc says `Sets the callback that will be called when the DOM has been fully constructed` but I wasn't sure

Comment (calebcordry) [reply src/utils/dom-writer.js]: eh I guess there are two cases, after transfer in normal case, or if `writer.close()` is called and there is no `parser.body` tried to handle both

File: src/utils/dom-writer.js [hunk 2]
```diff
@@ -103,7 +103,7 @@ export class DomWriterStreamer {
     /** @private {boolean} */
     this.eof_ = false;
 
-    /** @private {?Element} */
+    /** @private {?Promise<!Element>} */
```
Comment (kevinkimball) [inline src/utils/dom-writer.js]: Shouldn't this be something like `{!Promise<?Element>}`? Right now the element is optional but it appears you are making it required. Or are you only resolving the promise if there is an element?

Comment (dvoytenko) [reply src/utils/dom-writer.js]: I'd expect that body must always be available when the promise resolves.

**Comment by dvoytenko** [comment]
Overall, the PR looks ok, but I feel like otherwise it's a bit of a misuse of the API. It comes down to this question: do you ever need a separate part of the document at different times? E.g. do you need the `<head>` before the `layoutCallback`? Otherwise, you can simply buffer the response stream until it's needed.

**Comment by calebcordry** [comment]
My thought was that we could start our client side validation of the head as soon as possible, without being blocked by `layoutCallback`. This probably only makes a difference for ads far from initial viewport, but since we are introducing some additional logic it would be nice to keep it as fast as we can.

**Comment by dvoytenko** [comment]
> My thought was that we could start our client side validation of the head as soon as possible, without being blocked by `layoutCallback`. This probably only makes a difference for ads far from initial viewport, but since we are introducing some additional logic it would be nice to keep it as fast as we can.

There are many nuances here. In the meantime, do we keep the HTTP stream open? What kind of validation we will do? Only on the `head` element?

**Comment by calebcordry** [comment]
> In the meantime, do we keep the HTTP stream open? 

I thought we would keep the stream open until we run out of chunks? Any reason not to? Seems like we can just call `close()` at the end of response whether the body promise has resolved or not?

> What kind of validation we will do? Only on the `head` element?

Yes only the head is validated. Some more details [in this doc](https://docs.google.com/document/d/1UY3PkZmnxgx594-5e-RhZYtnlcKSo6ItgJOJxOBtPVo/edit#heading=h.iapm09vho0n8)

**Comment by calebcordry** [comment]
Closing in favor of #26945

**Pull request closed without merging**